Source/Tests/LinuxSyscalls/LinuxAllocator.cpp

```diff
@@ -0,0 +1,588 @@
+#include "Common/MathUtils.h"
+#include "Tests/LinuxSyscalls/LinuxAllocator.h"
+#include "Tests/LinuxSyscalls/Syscalls.h"
+
+#include <bitset>
+#include <map>
+#include <linux/mman.h>
+#include <unistd.h>
+#include <sys/mman.h>
+#include <sys/shm.h>
+
+namespace FEX::HLE {
+class MemAllocator32Bit final : public FEX::HLE::MemAllocator {
+private:
+  static constexpr uint64_t PAGE_SHIFT = 12;
+  static constexpr uint64_t PAGE_SIZE = 1 << PAGE_SHIFT;
+  static constexpr uint64_t PAGE_MASK = (1 << PAGE_SHIFT) - 1;
+  static constexpr uint64_t BASE_KEY = 16;
+  const uint64_t TOP_KEY = 0xFFFF'F000ULL >> PAGE_SHIFT;
+  const uint64_t TOP_KEY32BIT = 0x1F'F000ULL >> PAGE_SHIFT;
+
+public:
+  MemAllocator32Bit() {
+    // First 16 pages are taken by the Linux kernel
+    for (size_t i = 0; i < 16; ++i) {
+      MappedPages.set(i);
+    }
+    // Take the top page as well
+    MappedPages.set(TOP_KEY);
+    if (SearchDown) {
+      LastScanLocation = TOP_KEY;
+      LastKeyLocation = TOP_KEY;
+      LastKeyLocation32Bit = TOP_KEY32BIT;
+      FindPageRangePtr = &MemAllocator32Bit::FindPageRange_TopDown;
+    }
+    else {
+      LastScanLocation = BASE_KEY;
+      LastKeyLocation = BASE_KEY;
+      FindPageRangePtr = &MemAllocator32Bit::FindPageRange;
+    }
+  }
+  void *mmap(void *addr, size_t length, int prot, int flags, int fd, off_t offset) override;
+  int munmap(void *addr, size_t length) override;
+  void *mremap(void *old_address, size_t old_size, size_t new_size, int flags, void *new_address) override;
+  uint64_t shmat(int shmid, const void* shmaddr, int shmflg, uint32_t *ResultAddress) override;
+  uint64_t shmdt(const void* shmaddr) override;
+  static constexpr bool SearchDown = true;
+
+  // PageAddr is a page already shifted to page index
+  // PagesLength is the number of pages
+  void SetUsedPages(uint64_t PageAddr, size_t PagesLength) {
+    // Set the range as mapped
+    for (size_t i = 0; i < PagesLength; ++i) {
+      MappedPages.set(PageAddr + i);
+    }
+  }
+
+  // PageAddr is a page already shifted to page index
+  // PagesLength is the number of pages
+  void SetFreePages(uint64_t PageAddr, size_t PagesLength) {
+    // Set the range as unused
+    for (size_t i = 0; i < PagesLength; ++i) {
+      MappedPages.reset(PageAddr + i);
+    }
+  }
+
+private:
+  // Set that contains 4k mapped pages
+  // This is the full 32bit memory range
+  std::bitset<0x10'0000> MappedPages;
+  std::map<uint32_t, int> PageToShm{};
+  uint64_t LastScanLocation{};
+  uint64_t LastKeyLocation{};
+  uint64_t LastKeyLocation32Bit{};
+  std::mutex AllocMutex{};
+  uint64_t FindPageRange(uint64_t Start, size_t Pages);
```

Comment:
```suggestion
  uint64_t FindPageRange(uint64_t Start, size_t Pages) const;
```

Ditto for the `_TopDown` variant as well

Reply:
Resolved